File: backend/app/models/task.py
"""
Task model
"""

from sqlalchemy import Column, String, Text, DateTime, Integer, Numeric, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
from enum import Enum

from ..db.base import Base


class TaskStatus(str, Enum):
    """Task status enum"""
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    DONE = "done"
    CANCELLED = "cancelled"


class TaskPriority(str, Enum):
    """Task priority enum"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class Task(Base):
    """Task model"""

    __tablename__ = "tasks"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    document_id = Column(UUID(as_uuid=True), ForeignKey("documents.id", ondelete="SET NULL"), nullable=True, index=True)
    parent_task_id = Column(UUID(as_uuid=True), ForeignKey("tasks.id", ondelete="CASCADE"), nullable=True, index=True)

    # Content
    title = Column(String(255), nullable=False)
    description = Column(Text)

    # Scheduling
    due_date = Column(DateTime, index=True)
    estimated_duration_minutes = Column(Integer)

    # Status
    status = Column(String(50), default=TaskStatus.OPEN, index=True)
    priority = Column(String(50), default=TaskPriority.MEDIUM, index=True)

    # Metadata
    amount = Column(Numeric(10, 2))
    currency = Column(String(3), default="EUR")

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    completed_at = Column(DateTime)

    # Relationships
    user = relationship("User", back_populates="tasks")
    document = relationship("Document", back_populates="tasks")
    parent_task = relationship("Task", remote_side=[id], backref="subtasks")
    reminders = relationship("Reminder", back_populates="task", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Task {self.title} ({self.status})>"

File: frontend/lib/models/task.dart
class Task {
  final String id;
  final String title;
  final String? description;
  final DateTime? dueDate;
  final String status;
  final String priority;
  final double? amount;
  final String? currency;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? completedAt;

  Task({
    required this.id,
    required this.title,
    this.description,
    this.dueDate,
    required this.status,
    required this.priority,
    this.amount,
    this.currency,
    required this.createdAt,
    required this.updatedAt,
    this.completedAt,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      dueDate: json['due_date'] != null ? DateTime.parse(json['due_date']) : null,
      status: json['status'] as String? ?? 'open',
      priority: json['priority'] as String? ?? 'medium',
      amount: json['amount'] != null ? (json['amount'] as num).toDouble() : null,
      currency: json['currency'] as String?,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      completedAt: json['completed_at'] != null ? DateTime.parse(json['completed_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'due_date': dueDate?.toIso8601String(),
      'status': status,
      'priority': priority,
      'amount': amount,
      'currency': currency,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'completed_at': completedAt?.toIso8601String(),
    };
  }

  Task copyWith({
    String? title,
    String? description,
    DateTime? dueDate,
    String? status,
    String? priority,
    double? amount,
    String? currency,
  }) {
    return Task(
      id: id,
      title: title ?? this.title,
      description: description ?? this.description,
      dueDate: dueDate ?? this.dueDate,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
      completedAt: completedAt,
    );
  }

  bool get isOverdue {
    if (dueDate == null || status == 'done') return false;
    return dueDate!.isBefore(DateTime.now());
  }

  bool get isDone => status == 'done';
}
